design/Events_design.md

```diff
@@ -0,0 +1,113 @@
+
+This is a design proposal for how to handle events in Obscuro.
+
+## What are events in Ethereum
+
+### Smart contracts can emit events which are streamed from a node and generally used by UIs.
+
+For example:
+
+```emit Transfer(from, to, amount);```
+
+
+### A web app can subscribe to events by doing something like:
+
+```
+var subscription = web3.eth.subscribe('logs', {
+    address: '0x123456..', 
+    topics: ['0x12345...']
+}, function(error, result){
+    if (!error)
+        console.log(result);
+});
+- address - String|Array: An address or a list of addresses to only get logs from particular account(s).
+- topics - Array: An array of values which must each appear in the log entries. The order is important, if you want to leave topics out use null, e.g. [null, '0x00...']. You can also pass another array for each topic with options for that topic e.g. [null, ['option1', 'option2']]
+```
+
+A web app might request:
+- stream all "transfer" events where the 'to' field is my address
+- stream all "transfer" events from the USDC contract
+
+
+### How events work 
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is executed and events are emmitted, to match them against the filters requested by users and distribute them to the requester.(? - need to confirm this is 100% how it works) .
+
+*Note that there is no constraint on data access, since all data is public.*
+
+
+## Requirements and Problems in Obscuro
+
+- The high level requirement is to determine which account owner can view which events, and make sure, through encryption that nobody else can see those events.
+- Ideally we want to achieve this without any change to the API.
+
+
+#### Problem1: How do we authenticate event subscriptions? 
+We have a wallet extension with viewing key support, which we have to use ideally, to avoid creating another mechanism.
+
+#### Problem2: Assuming the enclave knows who is subscribed, how will the filtering work?
+
+E.g: Alice is requesting to listen to all transfer events. (this is a perfectly reasonable request on a public chain)
+
+There are 2 cases:
+1. The event was emitted by a smart contract as a result of executing a tx sent by Alice.
+2. The event was emitted as result of a Tx sent by Bob but it is somehow relevant to Alice.
+
+
+#### Problem 3: How to encrypt events before they leave the enclave
+
+#### Problem 4: The mechanism should not leak general information on what activity is happening.
+
+
+## Proposed solution
+
+### Prerequisites
+An event `E1` is emitted after executing transaction `TX1` with a from address `A1`.
+`E1` is formed of multiple topics, 2 of which are addresses (`A2` and `A3`), and there are some other random values.
+
+An event `E2` is emitted by a transaction `TX2` from an address `A4`.
+`E2` is formed of multiple topics, none of which are addresses.
+
+
+### Solution
+
+1. We modify the subscription call to accept a list of owning accounts. 
```

Comment:
We use "we" here in various places. Can you be explicit about what's done by the user, vs. what's done by the wallet extension? Hopefully this will help highlight what the impact on the user experience is (hopefully very little, it seems?).